### PR TITLE
fix(bruno-js): evaluate npm modules once instead of per script context (9.5 GB → 0.8 GB on a 2k-request run)

### LINKED ISSUE #9074: CLI 4.0.0: `bru run` retains ~13 MB per executed request (9.5 GB over a 2,170-request run; 3.0.3 peaks at 2 GB)
### I have checked the following:

- [x] I have searched existing issues and found nothing related to my issue. (Closest is #8137, which reports steady memory growth with CLI 3.4.x; this is 4.0.0 and an order of magnitude larger.)

### This bug is:

- [x] Making Bruno unusable for me.
- [x] This feature was working in a previous version but is broken in the current release.

### Bruno version

@usebruno/cli 4.0.0 (Node 24.12.0). Compared against @usebruno/cli 3.0.3 from the same lockfile.

### 💻 Operating System

Linux

### Specific OS Version / Distro

Ubuntu (dev machine, 64 GB RAM) and Ubuntu CI VMs (Azure Standard_D2ds_v6: 2 vCPU, 7.9 GB RAM, no swap)

### 🐞 Describe the bug

`bru run` with **@usebruno/cli 4.0.0** retains roughly **13 MB of memory per executed request** for the lifetime of the run. On a ~2,170-request run the runner process reaches **~6 GB after ~500 requests and ~9.5 GB by the end**. On an 8 GB CI VM the kernel OOM-kills the machine ~2 minutes into the run (the CI agent itself dies). **3.0.3 on the exact same collection peaks at ~2 GB** and finishes the run ~8× faster.

Measured by sampling the RSS of the runner process (`node …/@usebruno/cli/bin/bru.js`; note Node 24 names it `MainThread` in `ps`) every 5 s while running the same folder list:

| CLI | flags | peak RSS | duration | result |
|---|---|---|---|---|
| 4.0.0 | `--max-old-space-size=8192` | **9.5 GB** (6.1 GB at ~500 requests, 7.7 GB at ~1,000) | 514 s | 2170/2170 pass (64 GB machine) |
| 4.0.0 | `--max-old-space-size=2048` | 5.7 GB at 63 s, still climbing | — | cap has no effect → the growth is not in the V8 old space (external/native memory) |
| 4.0.0 | `--reporter-skip-response-body` | 2.4 GB at 182 requests, same slope | — | not the response bodies |
| 3.0.3 | `--max-old-space-size=8192` | **2.0 GB** | 65 s | 2170/2170 pass |

Additional data points:

- The generated `results.html` is only 15 MB and `results.xml` 2 MB, so the retained memory is not the reporter data.
- Growth is roughly linear in the number of executed requests (~13 MB each), regardless of response size (most responses here are a few KB).
- `--reporter-skip-response-body`, `--reporter-skip-headers`, and lowering `--max-old-space-size` all leave the slope unchanged.
- The same collection run at a much lower request rate (~2 req/s against a remote environment, ~20 min) completes on the same 8 GB VM, so the leak is per request, not per second.

Sample of the per-5 s RSS log for 4.0.0 (elapsed seconds, requests executed so far, runner RSS):

```
2s      0 requests    319 MB
52s   505 requests   6157 MB
103s  990 requests   7713 MB
153s 1503 requests   7724 MB
254s 2316 requests   9500 MB
405s 3598 requests   9509 MB   (count includes requests executed via bru.runRequest)
```

And 3.0.3 on the same run:

```
1s      0 requests    301 MB
16s   448 requests    667 MB
32s  1217 requests   1435 MB
62s  2175 requests   1867 MB
77s  2468 requests   1849 MB
```

### 📝 Steps to reproduce

1. A YAML (OpenCollection format) collection with ~4,400 requests in total; the run covers ~2,170 of them across 16 top-level folders. Requests use `auth: bearer` with environment variables, small JSON bodies, `assert` blocks, `set-variable` actions, and a number of `after-response` / `before-request` scripts. Setup helpers are exposed on the collection and invoked with `bru.runRequest(...)` from `before-request` scripts (~110 helper chains of ~15 requests each per run).
2. Run:

   ```
   NODE_OPTIONS=--max-old-space-size=8192 \
   npx @usebruno/cli@4.0.0 run -r --env Local "Folder A" "Folder B" … \
     --exclude-tags=unhappy \
     --reporter-skip-headers Authorization \
     --reporter-junit results.xml --reporter-html results.html \
     --sandbox=developer
   ```

3. In another shell, sample the runner every 5 s:

   ```
   while sleep 5; do ps -eo rss=,etimes=,args= | grep -E 'usebruno/cli' | grep -v grep | sort -rn | head -1; done
   ```

4. Observe RSS growing ~13 MB per executed request. Repeat with `npx @usebruno/cli@3.0.3` (same collection, same lockfile otherwise): RSS stays around 2 GB.

### 📁 Collection to reproduce

The collection is a private internal API suite that I cannot attach. Its relevant characteristics are described above; I'm happy to run any diagnostic build or flag against it and report back (e.g. a build that prints heap/external memory per request, or one with `bru.runRequest` result caching disabled).

### 📷 Screenshots/Live demo link

CI symptom on an 8 GB VM with 4.0.0 — host resource sample interleaved with the Bruno output, 60 s before the agent was OOM-killed (`MainThread` is the CLI runner):

```
[sample 17:27:03] cpu=83% steal=0% mem_used=5083MB avail=2853MB/7937MB top: MainThread=1434MB sqlservr=1219MB dotnet=291MB
[sample 17:27:33] cpu=87% steal=0% mem_used=6231MB avail=1706MB/7937MB top: MainThread=2470MB sqlservr=1148MB dotnet=340MB
[sample 17:28:04] cpu=86% steal=0% mem_used=7343MB avail=593MB/7937MB  top: MainThread=3550MB sqlservr=1015MB dotnet=323MB
<agent lost>
```

Ref: BRU-4445

### PR DESCRIPTION
**Pull request by dgyesbreghs**
## Description

Fixes #9074.
Ref - BRU-4445

The node-vm sandbox (`--sandbox=developer`) creates a fresh `vm` context for every script execution, and `cjs-loader` evaluated **npm modules inside that context**, cached only in the per-context `localModuleCache`. A collection-level script doing `require('@faker-js/faker')` (or `moment`, `nanoid`, ...) therefore re-evaluated the whole package on **every request**. Measured per script run: ~20 MB of heap plus ~8 MB of ArrayBuffers, in contexts V8 only reclaims under heap pressure. On a 2,170-request `bru run` the CLI reached **9.5 GB RSS** (6 GB after ~500 requests) and OOM-killed our 8 GB CI agents; `@usebruno/cli` 3.0.3 — which handed npm modules to the host `require` — peaked at 2 GB on the same collection.

This PR evaluates npm modules **once per process**, in a dedicated shared vm context, and shares their exports with every script, the way Node's own `require` cache behaves:

- `cjs-loader.js`: new `sharedNpmModuleCache` + `getSharedNpmContext()`. The shared context gets the same `safeGlobals`/typed arrays as script contexts; the Bruno objects a module may reference as globals (`bru`, `req`, `res`, `test`, `expect`, `assert`, `__brunoTestResults`, `__bruSetScope`, `jwt`, `console`, `scriptingConfig`) are defined as **accessors that resolve to the script context currently executing**, so a module evaluated during request 1 still sees request 42's `bru` when called from request 42 (the existing `should provide bru object to npm modules` test keeps passing, and a new test covers the cross-execution case).
- `index.js`: `enterScriptContext(scriptContext)` / `exitScriptContext()` around `runInContext` (a stack, so `bru.runRequest` nesting works).
- Collection-local modules (`./scripts/x.js`) are unchanged: still evaluated per script context via `localModuleCache` (a test pins that).

Behavioural note for reviewers: exports of npm modules are now objects from the shared context rather than from the script's own context. That is the pre-4.0 behaviour (host-context objects), so `instanceof` checks across the boundary behave as they did in 3.x.

## Measurements (same collection, 2,170 requests, Node 24)

| | peak RSS | duration |
|---|---|---|
| `main` (4.0.0) | 9.5 GB | 514 s |
| this PR | 818 MB | 205 s |
| 3.0.3 for reference | 2.0 GB | 65 s* |

\* 3.0.3 runs the suite faster for reasons unrelated to this change; the remaining gap is a separate topic.

## Tests

- `packages/bruno-js`: 638/638 (`npm test`), including 3 new tests in `node-vm/index.spec.js`: npm module evaluated once across executions, cached module sees the *current* script's `bru`, local modules stay per context.
- Verified against the real collection with `bru run` while sampling the runner's RSS every 5 s.
- `npx eslint` on the changed files: 0 errors.

Companion PR for the per-request socket retention in `bruno-cli` that this change exposes: #9079.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved consistency when using npm modules across multiple script executions by reusing shared module instances.
  * Ensured modules access the correct active script context during concurrent, nested, and asynchronous executions.
  * Preserved isolation for collection-local modules.
  * Improved support for modules that capture or destructure Bruno context values.

* **Tests**
  * Added coverage for shared module reuse, context handling, concurrent execution, and collection-local module behavior.

### CODE REVIEW

File: packages/bruno-js/src/sandbox/node-vm/cjs-loader.js [hunk 1]
```diff
@@ -4,6 +4,95 @@ const path = require('node:path');
...
+  for (const key of Object.keys(scriptContext)) {
+    if (key !== 'global' && key !== 'globalThis' && key !== 'require') {
+      defineDynamicGlobal(key);
+    }
+  }
+  activeScriptContexts.push(scriptContext);
+}
+
+function exitScriptContext() {
+  activeScriptContexts.pop();
```
[the remark this answers is not in the document]
Comment (dgyesbreghs) [reply packages/bruno-js/src/sandbox/node-vm/cjs-loader.js]: Agreed — the stack only survived LIFO completion. Pushed b4edf4c: the active script context is now tracked with `AsyncLocalStorage` (`runWithScriptContext(scriptContext, () => script.runInContext(...))`, accessors read `activeScriptContext.getStore()`), so interleaved executions each resolve their own `bru`/`req`/`res` whatever order they finish in, and nested `bru.runRequest` executions nest naturally. Added an `it.each` covering both completion orders; the *first-started finishes first* case fails against the previous stack implementation and passes now. bruno-js suite: 640/640.

File: packages/bruno-js/src/sandbox/node-vm/cjs-loader.js [hunk 2]
```diff
@@ -2,8 +2,103 @@ const vm = require('node:vm');
...
+
+function defineDynamicGlobal(key) {
+  if (Object.prototype.hasOwnProperty.call(sharedNpmSandbox, key)) {
+    return;
+  }
+  Object.defineProperty(sharedNpmSandbox, key, {
+    enumerable: true,
+    configurable: true,
+    get: () => activeScriptContext.getStore()?.[key]
+  });
```
[the remark this answers is not in the document]
Comment (dgyesbreghs) [reply packages/bruno-js/src/sandbox/node-vm/cjs-loader.js:160]: Right — a plain accessor only helps when `bru` is re-read. Pushed 915b63a: each Bruno global on the shared context is now a late-bound facade (`facadeFor(key)`, a Proxy over a callable target) whose traps (`get`/`set`/`has`/`ownKeys`/`getOwnPropertyDescriptor`/`getPrototypeOf`/`apply`) resolve the active script context on every access, and methods are handed out as late-bound wrappers so `const { getVar } = bru` at module scope also keeps calling the current script's `getVar`. Added a test where the module does `const captured = bru; const { getVar } = bru;` during execution A and is called from B — it read A's value before this change, B's now. 641/641.

**Comment by dgyesbreghs** [comment]
Re the `typeof` nitpick: agreed, a facade over a function target made `typeof console` / `typeof bru` report `'function'` inside npm modules. Pushed 0bbecd7: the facade target is now a plain object unless the global's current value is a function (`test`, `expect`, `assert`, `__bruSetScope` as the fallback set when a facade is created outside an execution), so `typeof bru === 'object'` and `typeof test === 'function'` again — asserted in the capture test (`object/object/function`). 641/641, lint clean.

File: packages/bruno-js/src/sandbox/node-vm/cjs-loader.js [hunk 3]
```diff
@@ -2,8 +2,191 @@ const vm = require('node:vm');
...
+ */
+function facadeFor(key) {
+  if (facades.has(key)) {
+    return facades.get(key);
+  }
+  const current = () => activeScriptContext.getStore()?.[key];
+  const isMissing = (value) => value === undefined || value === null;
+  const initial = current();
+  const callable = isMissing(initial) ? CALLABLE_CONTEXT_KEYS.has(key) : typeof initial === 'function';
+  const target = callable ? () => {} : {};
```
[the remark this answers is not in the document]
Comment (dgyesbreghs) [reply packages/bruno-js/src/sandbox/node-vm/cjs-loader.js]: Agreed. Pushed d4a09a1: callability is now decided per access from the value the *current* execution provides, with one cached facade per (key, callable) pair — so a key that is missing or an object in one execution and a function in another is served correctly in both, and the fallback set is gone. While at it: a key the current execution does not provide now reads as plain `undefined` inside npm modules (as it would in the script's own context, e.g. `typeof res === 'undefined'` before a response) instead of an empty facade. New test: `helper` undefined in execution A (module sees `'undefined'`), a function in execution B (module calls it). 642/642, lint clean.

**Comment by helloanoop** [comment]
Thanks for the PR @dgyesbreghs 
Our team is looking into this.

**Comment by dgyesbreghs** [comment]
Thanks @helloanoop! Happy to adjust anything. FYI the branch (and #9079's) is being kept rebased on `main` continuously, and the reproduction numbers in the description are from a real 2,170-request collection, so I can quickly re-measure any alternative approach your team prefers.

**Comment by dgyesbreghs** [comment]
On the red *Playwright E2E (Linux) 4/4* shard: the single failure is `variables-tab.spec.ts:453 › restores the scroll position after leaving and reopening the tab`, alongside 6 flaky scroll/shortcut/websocket-UI tests in the same shard — all DOM/scroll behaviours with no overlap with this PR's sandbox/module-loading change (350 others in the shard passed, and unit/CLI/SSL/OAuth suites are green on both platforms). Looks like shard flakiness; a re-run should clear it.

**Comment by dgyesbreghs** [comment]
Same triage for the now-finished *Playwright E2E (Windows)* job (19 failed / 18 flaky / 1383 passed, all failures 30 s timeouts): the failing specs (`snapshots/*` persistence, `variables-tab`, `environment-tabs`, websocket scroll) are the same families that fail on `main`'s own Windows Tests run for a docs-only change (run 33496659277), and overlap with the flaky set addressed by #9122 today — so this looks like the suite's baseline Windows instability rather than this PR. The suites that exercise the change directly (bruno-js unit, CLI on both platforms) are green.

**Comment by sid-bruno** [comment]
@dgyesbreghs 

Have merged your approach and after a few additional tests and cases decided to go into a different direction, this also has some edge cases that would fail but depending on whether the trade offs make sense between the 2, we'll either go back to your approach(and patch a few issues that'll need a hacky run around) or keep the one we've taken (which will have some async gaps since I'm no longer maintaining context with ALS). 

For now, if you could help me with checking https://github.com/usebruno/bruno/pull/9170 with your 2,170 request collection would help me figure out any possible behaviors that we might have missed.
